# Evolutionary distances and summary statistics calculation

These scripts calculate summary statistics and evolutionary distances between pseudogene-gene pairs (`summary_pseudogenes.sh` and `summary_pseudogenes.py`) and anchorpair genes (`summary_anchorpairs.sh` and `summary_anchorpairs.py`).

In order to get insight into the evolutionary distance between a pseudogene and a gene, we calculated the KS (number of synonymous substitutions per synonymous site), KA (number of non-synonymous substitutions per non-synonymous site) and KA/KS (proportion of non-synonymous to synonymous substitutions). To do this, the sequences of the pseudogene ex-ons were obtained from the genome using the coordinates and combined. Subsequently, the pseudogene sequence was aligned to the functional coding gene using MACSE v2.06 (Ran-wez et al., 2018). MACSE can accommodate pseudogenes by recognizing them as less relia-ble sequences, assigning them a reduced penalty for frameshifts and stop codons in contrast to functional genes. Next, the obtained alignment was converted from FASTA to AXT format, and evolutionary distances were calculated using KaKs_Calculator 2.0 (Wang et al., 2010) through model averaging (MA), considering various evolutionary models. In addition, evolu-tionary distances were calculated between anchorpair genes using the same programs.
